### PR TITLE
feat: supported Gemini Pro 1.5

### LINKED ISSUE #63: Support Gemini 1.5

### PR DESCRIPTION
**Pull request by adubovik**
Resolves #63 

As per [doc](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/send-multimodal-prompts):
* Gemini Pro 1.0 Vision supported attachments:
  * Video
  * PDF
* Gemini Pro 1.5 supported attachments:
  * Images
  * Audio
  * Video
  * PDF

### CODE REVIEW

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by adubovik** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-adapter-vertexai-pr-75.nightly-test.deltixhub.io
>E2E tests status: failed

File: aidial_adapter_vertexai/chat/gemini/adapter.py [hunk 1]
```diff
@@ -24,24 +24,38 @@
...
+HarmBlockThreshold = generative_models.HarmBlockThreshold
+GenFinishReason = generative_models.FinishReason
 
-BLOCK_NONE_SAFETY_SETTINGS: Dict[HarmCategory, HarmBlockThreshold] = {
-    HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,
-    HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,
-    HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,
-    HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,
+default_safety_settings: Dict[HarmCategory, HarmBlockThreshold] = {
+    HarmCategory.HARM_CATEGORY_UNSPECIFIED: HarmBlockThreshold.BLOCK_ONLY_HIGH,
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/chat/gemini/adapter.py:54]: Why not NONE?

File: aidial_adapter_vertexai/chat/gemini/inputs.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        # Placing Images/Video before the text as per
+        # https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/send-multimodal-prompts?authuser=1#image_best_practices
+        parts.append(Part.from_text(self.content))
+
+        return parts
+
+    def to_content(self) -> Content:
+        return Content(
+            role=get_part_role(self.message.role),
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/chat/gemini/inputs.py]: Maybe to_gemini_role or from_dial_role?

File: aidial_adapter_vertexai/chat/gemini/inputs.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+def derive_attachment_mime_type(attachment: Attachment) -> Optional[str]:
+    type = attachment.type
+
+    if type is None:
+        # No type is provided. Trying to guess the type from the Data URL
+        if attachment.url is not None:
+            resource = Resource.from_data_url(attachment.url)
+            if resource is not None:
+                return resource.mime_type
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/chat/gemini/inputs.py:64]: May it be octet-stream here?

File: aidial_adapter_vertexai/chat/gemini/processor.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    init_validator: InitValidator | None = None
+    post_validator: PostValidator | None = None
+
+    @property
+    def mime_types(self) -> List[str]:
+        return list(self.file_types.keys())
+
+    @property
+    def file_exts(self) -> List[str]:
+        def flatten(value: Union[str, List[str]]):
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/chat/gemini/processor.py]: Is it an outdated name for the function? It's minor but looks a bit misleading.

File: aidial_adapter_vertexai/chat/gemini/processor.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            return value if isinstance(value, list) else [value]
+
+        return [
+            ext for exts in self.file_types.values() for ext in flatten(exts)
+        ]
+
+    async def process(
+        self, file_storage: Optional[FileStorage], attachment: Attachment
+    ) -> Optional[Resource | str]:
```
Comment (Oleksii-Klimov) [inline aidial_adapter_vertexai/chat/gemini/processor.py:58]: | None

**Review by Oleksii-Klimov — DISMISSED** [review]

File: aidial_adapter_vertexai/chat/gemini/adapter.py (hunk 1 above)
Comment (adubovik) [reply aidial_adapter_vertexai/chat/gemini/adapter.py:54]: For some reason it `BLOCK_NONE` doesn't work for certain use cases.

I have an audio-file, which triggers `400 Request contains an invalid argument` error when I use `BLOCK_NONE`.

File: aidial_adapter_vertexai/chat/gemini/inputs.py (hunk 2 above)
Comment (adubovik) [reply aidial_adapter_vertexai/chat/gemini/inputs.py]: Renamed to `from_dial_role`

File: aidial_adapter_vertexai/chat/gemini/inputs.py (hunk 3 above)
Comment (adubovik) [reply aidial_adapter_vertexai/chat/gemini/inputs.py:64]: Yes, if data URL specified octet-stream as its media type. But I don't know what could we do in such a case.
I refactored the method.

File: aidial_adapter_vertexai/chat/gemini/processor.py (hunk 4 above)
Comment (adubovik) [reply aidial_adapter_vertexai/chat/gemini/processor.py]: Renamed to `to_list`.

**Review by Oleksii-Klimov — APPROVED** [review]

**Pull request merged**